New FP Parsing:  Use relevant part of a substring

tbkka opened the pull request:
The new FP parsing code inadvertently used the entire backing string, even when the substring referred to just a portion.

This addresses a failure in the extended `UnicodeScalarProperties.swift` test, which parses an input file by reading a line into a String, then extracting Substrings.  (rdar://166258748)

tbkka committed:
Use the correct portion of a substring

Catfish-Man commented on stdlib/public/core/FloatingPointParsing.swift.gyb:
```diff
@@ -256,15 +256,21 @@ extension ${Self}: LosslessStringConvertible {
+    } else {
+      // Otherwise, make a copy so we have contiguous UTF8...
+      let string = String(text)
+      parsed = unsafe string._guts.withFastUTF8 { chars -> ${Self}? in
```
Why not just use https://developer.apple.com/documentation/swift/string/makecontiguousutf8() ?

glessard replied:
The current implementation does always produce a string that supports `withFastUTF8`, but we might as well be sure and call `makeContiguousUTF8()`. I would suggest `withUTF8`, since it is a wrapper for `makeContiguousUTF8()` followed by `_guts.withFastUTF8`…

glessard commented:
> In particular, I've not made any effort to test cases where a string is not "fast UTF8" capable; advice appreciated.

There are examples of tests of bridged substrings in test/stdlib/Span/BridgedStringUTF8ViewSpanTests.swift

glessard commented on stdlib/public/core/FloatingPointParsing.swift.gyb:
```diff
@@ -256,15 +256,21 @@ extension ${Self}: LosslessStringConvertible {
   @available(SwiftStdlib 5.3, *)
   public init?(_ text: Substring) {
-    // TODO: Someday, this whole function should simplify down to just:
-    // ${Self}(text.utf8.span)
```
Fair. It would be nice to show what it looks like when no availability workarounds are needed, though.

glessard commented on stdlib/public/core/FloatingPointParsing.swift.gyb:
```diff
@@ -256,15 +256,21 @@ extension ${Self}: LosslessStringConvertible {
+    } else {
+      // Otherwise, make a copy so we have contiguous UTF8...
+      let string = String(text)
+      parsed = unsafe string._guts.withFastUTF8 { chars -> ${Self}? in
```
```suggestion
      var string = String(text)
      parsed = string.withUTF8 { chars -> ${Self}? in
```
If `text` is a substring of a bridged string that has the same range of indices as its base string (e.g. `let text = bridgedString[...]`), then the return string will still be a bridged string, and that wouldn't support `withFastUTF8`.

tbkka committed:
Test short and long bridged strings

Simplify the top-level initializer by using `Substring.withUTF8`
to get a correct span over contiguous UTF8.

tbkka commented on test/stdlib/ParseFloat64.swift:
```diff
@@ -328,6 +328,14 @@ tests.test("Decimal Floats") {
+  expectNotNil(parsed)
+  expectEqual(parsed!.bitPattern, (2.0).bitPattern)
+}
+
```
I added two tests that exercise substrings of bridged strings:  One is short enough to use tagged/inline strings; the other is long enough to prevent same.

tbkka committed:
Comment out Bridged tests that cannot run on Linux

tbkka commented:
@swift-ci Please test

tbkka committed:
Break out bridging to a separate test source file

tbkka commented:
@swift-ci Please test

MaxDesiatov commented:
@swift-ci test windows

glessard reviewed (approved):
Looks good, thanks!
